Clean links refactor

herregroen opened the pull request:
## Context
<!--
What do we want to achieve with this PR? Why did we write this code?
-->

* The current links architecture seems to be heavily overengineered. This branch should bring all the link table logic to the `src` directory reusing the concepts we already use there.

## Todo

- [x] Cleanup the dependency chain, merge / remove unneeded files
- [x] Restructure / rename existing classes to make the concepts clear.
- [x] Refactor the storage and installer classes to repositories and migrations.
- [x] Refactor the accessible classes to conditionals.
- [x] Refactor the indexing functionality (indexing plus query classes) to use the indexing framework.
- [x] Everything with register hooks becomes an integration.
- [x] Make sure the feature toggle is no longer a toggle for the backend functionality but only for the features built on top of it.
- [x] Create many to one relationship between links and indexables.
- [x] Write unit tests for new classes.

## Summary

<!--
Attach one of the following labels to the PR: `changelog: bugfix`, `changelog: enhancement`, `changelog: other`, `changelog: non-user-facing`.
If the changelog item is a bugfix, please use the following sentence structure: Fixes a bug where ... would ... (when ...).
If the changelog item is meant for the changelog of another repo, start you changelog item with the repo name between square brackets, for example: * [wordpress-seo-premium] Fixes a bug where ....
If the same changelog item is applicable to multiple changelogs/repos, add a separate changelog item for all of them.
-->
This PR can be summarized in the following changelog entry:

* Refactors the internal link indexing to modern standards.
* Adds indexing of images to the link indexing.

## Test instructions
<!--
Please follow these guidelines when creating test instructions:
- Please provide step-by-step instructions how to reproduce the issue, if applicable.
- Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->
This PR can be tested by following these steps:

* Set the `link_count` column of all indexables to `NULL`
* Make sure you have a number of posts that include a variety of:
  - Internal links ( a elements with a href that points to the same domain )
  - External links ( a elements with a href to another domain )
  - Internal images ( img element with a src that points to the same domain )
  - External images ( img element with a src that points to another domain )
* Run the link indexation tool ( if available, if you made the above posts on this PR it may not be as they would already be indexed )
* You should now see all links, both normal and images, reflected in the `seo_links` table.
* Update a post, changing which links are present.
* These changes should be reflected in the `seo_links` table after saving.
* Check the link count columns in the post overview, they should be accurate for all normal links ( not images ).
* Navigate to the next page, the links should still be accurate.


## UI changes
* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation
* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Takes care of a lot of https://github.com/Yoast/wordpress-seo/issues/15161

omarreiss commented on src/integrations/admin/link-count-columns-integration.php:
```diff
@@ -25,75 +38,137 @@ class WPSEO_Link_Columns {
 
 	/**
-	 * Register hooks that require to be registered after `init`.
+	 * Register hooks that need to be registered after `init`.
```
Maybe also note why this is?

omarreiss commented on src/integrations/admin/link-count-tools-integration.php:
```diff
@@ -0,0 +1,171 @@
+		$data = [
+			'amount'  => $this->get_total_unindexed(),
+			'ids'     => [
+				'modal'        => '#yoast-link-indexing-modal',
```
Are these id's used anywhere else in the PHP?

herregroen replied:
Here and in the presenter. Due to the number of them I felt constants were a bit overkill.

With internal linking finalized it may be worth seeing if we can abstract all these indexing modals into one class so these IDs can get a uniform format which would allow us to abstract them here as well.

Not sure that should be part of this PR though.

omarreiss commented on src/integrations/admin/link-count-tools-integration.php:
```diff
@@ -0,0 +1,171 @@
+				'nonce'     => \wp_create_nonce( 'wp_rest' ),
+			],
+			'message' => [
+				'indexingCompleted' => '<span class="yoast-check">' . \esc_html__( 'Good job! All the links in your texts have been counted.', 'wordpress-seo' ) . '</span>',
```
Any reason why these translations are not JS territory?

herregroen replied:
This same settings object is passed for creating indexables, counting links and gathering prominent words. Having these here avoids out-of-context switch statements in the generic Javascript that handles all 3 of these.

omarreiss commented on src/builders/indexable-link-builder.php:
```diff
@@ -0,0 +1,342 @@
+
+		if ( empty( $links ) && empty( $images ) ) {
+			$indexable->link_count = 0;
+			$indexable->save();
```
@herregroen I see you left this. Can you address why?

omarreiss commented on src/builders/indexable-link-builder.php:
```diff
@@ -0,0 +1,342 @@
+			return [];
+		}
+
+		// TODO: Consider if the indexable should instead be the input?
```
@herregroen looks like you left this in. Needed for future consideration?

omarreiss replied:
Seems like I looked at an outdated diff.

omarreiss commented on src/conditionals/admin/posts-overview-and-ajax-conditional.php:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Conditional that is only met when in the admin.
+ */
+class Posts_Overview_And_Ajax_Conditional implements Conditional {
```
@herregroen any reason why you left this in?

omarreiss replied:
Huh weird, why is Github showing me this outdated diff....

omarreiss commented:
Ok, just one comment left about the seemingly redundant save. Plus unit the build is failing miserably. Almost there!

omarreiss commented:
CR 👍

johannadevos commented on src/actions/indexation/post-link-indexing-action.php:
```diff
@@ -0,0 +1,93 @@
+	 *
+	 * @var string
+	 */
+	const UNINDEXED_COUNT_TRANSIENT = 'wpseo-unindexed-post-link-count';
```
This is a detail, but for all the other transients we use underscores.

johannadevos commented:
Acceptance: 🚧 

- [ ] The styling of the buttons for the indexable indexation and the text link counter is very different.
- [ ] From a product point of view (but that may not be the scope of this issue), one or two lines with an explanation about the 'why' of the text link counter would be nice, as it is currently done for the indexables as well.
<img width="565" alt="Screenshot 2020-07-30 at 15 56 09" src="https://user-images.githubusercontent.com/20280513/88931559-45ae0980-d27d-11ea-9827-2c46d0419904.png">

- [ ] When the link counting has successfully completed, I get the green success alert at the top of my screen ("Good job! ...") However, at the bottom I still see the "Count links in your text" button. I can click it, it opens a modal that says 0/147 processed, and then quickly closes again. Only after a page refresh, I see the "Good job! ..." text at the bottom.
- [ ] The text link counter progress bar doesn't reach its maximum number. For me, it always stops at 119/147.

- [ ] I'm missing the green checkmark after a successful link count (like for the indexables).
<img width="569" alt="Screenshot 2020-07-31 at 10 11 43" src="https://user-images.githubusercontent.com/20280513/89014723-3dee7380-d316-11ea-8528-ff96f09e7235.png">

- [ ] Terms show some unexpected (at least for me?) behaviour. Most categories (but not all), and only a minority of tags, will get a value in `link_count` upon indexing. Other categories and tags will have `NULL`. I didn't spot an obvious difference between the terms that do and that don't get values for `link_count`. Terms without a value will get a value if they are edited and updated, but will have `NULL` again after a reset and reindex.

- [ ] `height`, `width` and `size` for images are `NULL` for all rows (I do have posts with links to images, both from the media library and externally).
- [ ] Not sure if I tested `language` and `region` the right way. But I created a post with a link to `https://disney.nl`, a language with an hreflang implementation. I did see that link in the `seo_links` table, but `language` and `region` were `NULL`.

Other:
- [x] Updating a post with links results in new entries in the `seo_links` table. 
- [x] `link_count` in the indexables table reflects the number of internal links.
- [x] The link count columns in the post overview look good.

johannadevos commented:
Tip: to reset everything for testing, run `UPDATE wp_yoast_indexable SET link_count = NULL; TRUNCATE wp_yoast_seo_links;`.

herregroen commented:
@johannadevos I've adressed most of your results.

Regarding the lack of an explanation test, we're planning to soon consolidate all indexing processes into one process so this will be fixed at that time, especially given that such a text was not present before and as such this is not a regression.

I wasn't able to reproduce the issue with the number of incoming links being wrong, could you please check to ensure there's no other post linking to the post you tested with?

hansjovis commented:
Acceptance test: 🚧

- [ ] The incoming link counts are wrong in my environment as well.
  - I have an explanation (at least in my case): redirects. When redirecting a post to another, there is a mismatch between the anchor link in the contents of a post, and the permalink of the linked post. So, without going through the redirects this is going to be difficult to solve, I think. 
- [ ] The image height and width in the `seo_links` table is always 150 by 150.
  - [`wp_get_attachment_image_src` standardly retrieves the image thumbnail](https://developer.wordpress.org/reference/functions/wp_get_attachment_image_src/), which is presumably always 150 by 150.

hansjovis commented:
Just asked product about the questions above.
* The incoming link counts / mismatched links will be fixed in a later iteration.
* The image width and height should be of the original uploaded image.

hansjovis commented:
I decided to fix the image width and height myself, since [it was an easy fix](https://github.com/Yoast/wordpress-seo/pull/15609/commits/7ee89a8d0521a0f1eb6cc2fc15e282075a506349).

Unfortunately, while testing the code I discovered a bug. Activating the plugin gives a fatal error:
```
Fatal error: Uncaught Error: Call to undefined function YoastSEO() in /var/www/html/wp-content/plugins/wordpress-seo/inc/class-post-type.php on line 88

Error: Call to undefined function YoastSEO() in /var/www/html/wp-content/plugins/wordpress-seo/inc/class-post-type.php on line 88

Call Stack:
    1.3331    6908856   1. QM_Collector_PHP_Errors->exception_handler() /var/www/html/wp-content/plugins/query-monitor/collectors/php_errors.php:0
```
I think that the DI container is called in the 'old' code before the container has been initialized. [It seems to be related to this commit](https://github.com/Yoast/wordpress-seo/pull/15609/commits/b2634932daa2e9a8f17a355d96c47ff40d60294d).

I am giving this back to you @herregroen, since I am missing the context to solve this myself.

herregroen commented:
@hansjovis I've added a commit that should fix the issue you encountered.

Thanks for fixing the image as well! That commit looks good!

hansjovis commented:
**Note**: The latest commit does fix the issue of the fatal error being thrown when activating the plugin.

herregroen commented:
@hansjovis Should be fixed now. Adding the functions to the composer.json caused issues in vendor prefixing our dependencies. I've now moved them to the top of the wpseo-main.php. I've also removed the PHP version check as we've been 5.6 minimum for quite a while now.

hansjovis commented:
Acceptance test: ✅ 

Fatal error on plugin activation is solved. I followed the test instructions again, just in case, and everything works as expected. The only thing that is expected not to work is the mismatched link counts, but that will be fixed in a later iteration.
